crates/turbo-tasks-memory/src/task.rs

```diff
@@ -620,16 +898,96 @@ impl Task {
         schedule_task
     }
 
+    fn scopes_dirty_or_active<const LIKELY_ACTIVE: bool, const INCREMENT_UNFINISHED: bool>(
+        &self,
+        scopes: &TaskScopes,
+        backend: &MemoryBackend,
+    ) -> bool {
+        if !LIKELY_ACTIVE && INCREMENT_UNFINISHED {
+            let mut active = false;
+            let mut processed_scopes = 0;
+            for scope in scopes.iter() {
+                backend.with_scope(scope, |scope| {
+                    scope.increment_unfinished_tasks(backend);
+                    log_scope_update!("add unfinished task: {} -> {}", *scope.id, *self.id);
+                    if !active {
+                        processed_scopes += 1;
+                        let mut state = scope.state.lock();
+                        if state.is_active() {
+                            active = true;
+                        } else {
+                            state.add_dirty_task(self.id);
+                        }
+                    }
+                })
+            }
+            if active {
+                for scope in scopes.iter().take(processed_scopes) {
+                    backend.with_scope(scope, |scope| {
+                        let mut state = scope.state.lock();
+                        if !state.is_active() {
+                            state.remove_dirty_task(self.id);
+                        }
+                    })
+                }
+            }
+            return active;
+        }
+        if LIKELY_ACTIVE {
+            if INCREMENT_UNFINISHED {
+                let mut active = false;
+                for scope in scopes.iter() {
+                    backend.with_scope(scope, |scope| {
+                        scope.increment_unfinished_tasks(backend);
+                        active = active || scope.state.lock().is_active();
+                    })
+                }
+                if active {
+                    return true;
+                }
+            } else {
+                if scopes
+                    .iter()
+                    .any(|scope| backend.with_scope(scope, |scope| scope.state.lock().is_active()))
+                {
+                    return true;
+                }
+            }
+        }
+        let mut processed_scopes = 0;
+        for scope in scopes.iter() {
```

Comment:
This could be

```suggestion
        for (processed_scopes, scope) in scopes.iter().enumerate() {
```

Although it might make it a bit more confusing.

Reply:
I like it